File: x3270/Common/Win32/mkversion.c
/*
 * Copyright (c) 2008-2009, Paul Mattes.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the names of Paul Mattes nor the names of his contributors
 *       may be used to endorse or promote products derived from this software
 *       without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY PAUL MATTES "AS IS" AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
 * EVENT SHALL PAUL MATTES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * This program is a C-language encapsulation of all but the last line of:
 *
 *	#! /bin/sh
 *	# Create version.o from version.txt
 *
 *	# Ensure that 'date' emits 7-bit U.S. ASCII.
 *	LANG=C
 *	LC_ALL=C
 *	export LANG LC_ALL
 *
 *	set -e
 *
 *	. ./version.txt
 *	builddate=`date`
 *	sccsdate=`date +%Y/%m/%d`
 *	user=${LOGNAME-$USER}
 *
 *	# Create an all numeric timestamp for rpqnames.
 *	# rpq.c will return this string of numbers in bcd format
 *	# It is OK to change the length (+ or -), but use
 *	# decimal (0-9) digits only. Length must be even number of digits.
 *	rpq_timestamp=`date +%Y%m%d%H%M%S`
 *
 *	trap 'rm -f version.c' 0 1 2 15
 *
 *	cat <<EOF >version.c
 *	char *build = "${2-x3270} v$version $builddate $user";
 *	char *app_defaults_version = "$adversion";
 *	static char sccsid[] = "@(#)${2-x3270} v$version $sccsdate $user";
 *
 *	const char *build_rpq_timestamp = "$rpq_timestamp";
 *	const char *build_rpq_version = "$version";
 *	EOF
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static char *
NewString(char *s)
{
	char *t = malloc(strlen(s) + 1);

	if (t == NULL) {
		fprintf(stderr, "out of memory\n");
		exit(1);
	}
	return strcpy(t, s);
}

int
main(int argc, char *argv[])
{
	FILE *f;
	char buf[1024];
	char *version = NULL;
	char *adversion = NULL;
	char *user;
	__time64_t t;
	char *builddate;
	struct tm *tm;
	char sccsdate[128];
	char rpqtime[128];
	int is_w = 0;
	char *ofile = "version.c";
	char *progname = "wc3270";

	if (argc > 1 && !strcmp(argv[1], "-w")) {
		is_w = 1;
		ofile = "wversion.c";
		argv++;
		argc--;
	}
	if (argc > 1)
		progname = argv[1];

	/* Read up version.txt. */
	f = fopen("version.txt", "r");
	if (f == NULL) {
		perror("version.txt");
		return 1;
	}
	while (fgets(buf, sizeof(buf), f) != NULL) {
		if (!strncmp(buf, "version=\"", 9)) {
			char *q;

			version = NewString(buf + 9);
			q = strchr(version, '"');
			if (q == NULL) {
				fprintf(stderr,
					"syntax error in version.txt\n");
				return 1;
			}
			*q = '\0';
		} else if (!strncmp(buf, "adversion=\"", 11)) {
			char *q;

			adversion = NewString(buf + 11);
			q = strchr(adversion, '"');
			if (q == NULL) {
				fprintf(stderr,
					"syntax error in version.txt\n");
				return 1;
			}
			*q = '\0';
		}
	}
	fclose(f);
	if (version == NULL || adversion == NULL) {
		fprintf(stderr,
			"missing version= or adversion= in version.txt\n");
		return 1;
	}

	/* Grab the username. */
	user = getenv("USERNAME");
	if (user == NULL) {
		fprintf(stderr, "No %USERNAME%?\n");
		return 1;
	}

	/* Format the dates. */
	_time64(&t);
	builddate = NewString(_ctime64(&t));
	builddate[strlen(builddate) - 1] = '\0';
	tm = _localtime64(&t);
	sprintf(sccsdate, "%d/%02d/%02d",
		tm->tm_year + 1900,
		tm->tm_mon + 1,
		tm->tm_mday);
	sprintf(rpqtime, "%02d%02d%02d%02d%02d",
		tm->tm_year + 1900,
		tm->tm_mon + 1,
		tm->tm_mday,
		tm->tm_hour,
		tm->tm_min,
		tm->tm_sec);

	/* Create the code. */
	f = fopen(ofile, "w");
	if (f == NULL) {
		perror(ofile);
		return 1;
	}
	if (is_w) {
		fprintf(f, "char *wversion = \"%s\";\n", version);
	} else {
		fprintf(f, "char *build = \"%s v%s %s %s\";\n",
			progname, version, builddate, user);
		fprintf(f, "char *app_defaults_version = \"%s\";\n",
			adversion);
		fprintf(f, "static char sccsid[] = \"@(#)%s v%s %s %s\";\n",
			progname, version, sccsdate, user);
		fprintf(f, "const char *build_rpq_timestamp = \"%s\";\n",
			rpqtime);
		fprintf(f, "const char *build_rpq_version = \"%s\";\n",
			version);
	}
	fclose(f);

	return 0;
}

File: Common/mkversion.py
#!/usr/bin/env python3
#
# Copyright (c) 2025 Paul Mattes.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#     * Neither the names of Paul Mattes nor the names of his contributors
#       may be used to endorse or promote products derived from this software
#       without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY PAUL MATTES "AS IS" AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
# EVENT SHALL PAUL MATTES BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
# OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Create version.c from version.txt
#  mkversion [-o outfile] [program [path-to-version.txt]]
#
# Formerly a shell script:
#
# Ensure that 'date' emits 7-bit U.S. ASCII.
# LANG=C
# LC_ALL=C
# export LANG LC_ALL
# 
# set -e
# 
# if date --help >/dev/null 2>&1
# then    # GNU date
#         ddopt="-d@"
# else    # BSD date
#         ddopt="-r"
# fi
# 
# . ${2-./version.txt}
# date="date -u"
# [ -n "$SOURCE_DATE_EPOCH" ] && date="$date $ddopt$SOURCE_DATE_EPOCH"
# builddate=`$date`
# sccsdate=`$date +%Y/%m/%d`
# user=${LOGNAME-$USER}
# 
# # Create an all numeric timestamp for rpqnames.
# # rpq.c will return this string of numbers in bcd format
# # It is OK to change the length (+ or -), but use
# # decimal (0-9) digits only. Length must be even number of digits.
# rpq_timestamp=`$date +%Y%m%d%H%M%S`
# 
# app=${1-x3270}
# cat <<EOF
# const char *app = "$app";
# const char *build = "$app v$version $builddate $user";
# const char *app_defaults_version = "$adversion";
# const char *cyear = "$cyear";
# const char sccsid[] = "@(#)$app v$version $sccsdate $user";
# 
# const char *build_rpq_timestamp = "$rpq_timestamp";
# const char *build_rpq_version = "$version";
# EOF

from datetime import datetime, timezone
import locale
import os
import re
import sys

def mkversion(outfile_name: str, program: str, version_txt: str):
    # Set the locale.
    locale.setlocale(locale.LC_ALL, 'C')

    # Read in version.txt.
    ver = {}
    lno = 0
    infile = open(version_txt)
    while True:
        line = infile.readline()
        if line == '':
            break
        lno += 1
        line = line.strip()
        m = re.fullmatch('^([a-z]+)="(.*)"$', line)
        if m == None:
            print(f'Syntax error in version.txt at line {lno}', file = sys.stderr)
            continue
        ver[m.group(1)] = m.group(2)
    infile.close()

    if not 'version' in ver or not 'adversion' in ver or not 'cyear' in ver:
        print('Missing fields(s) in version.txt', file = sys.stderr)
        exit(1)
    version = ver['version']
    adversion = ver['adversion']
    cyear = ver['cyear']

    # Open the output file.
    if outfile_name != None:
        outfile = open(outfile_name, 'w')
    else:
        outfile = sys.stderr

    # Get the date.
    if 'SOURCE_DATE_EPOCH' in os.environ:
        now = datetime.strptime(os.environ['SOURCE_DATE_EPOCH'], '%a %b %d %H:%M:%S %Z %Y').replace(tzinfo=timezone.utc)
    else:
        now = datetime.now(timezone.utc)
    builddate = datetime.strftime(now, '%a %b %d %H:%M:%S %Z %Y')
    sccsdate = datetime.strftime(now, '%Y/%m/%d')
    rpq_timestamp = datetime.strftime(now, '%Y%m%d%H%M%S')

    # Get the user.
    if 'USER' in os.environ:
        user = os.environ['USER']
    elif 'USERNAME' in os.environ:
        user = os.environ['USERNAME']
    elif 'LOGNAME' in os.environ:
        user = os.environ['LOGNAME']
    else:
        user = 'unknown user'

    # Print it all out.
    print(f'const char *app = "{program}";', file=outfile)
    print(f'const char *build = "{program} v{version} {builddate} {user}";', file=outfile)
    print(f'const char *app_defaults_version = "{adversion}";', file=outfile)
    print(f'const char *cyear = "{cyear}";', file=outfile)
    print(f'const char sccsid[] = "@(#){program} v{version} {sccsdate} {user}";', file=outfile)
    print(file=outfile)
    print(f'const char *build_rpq_timestamp = "{rpq_timestamp}";', file=outfile)
    print(f'const char *build_rpq_version = "{version}";', file=outfile)

    if outfile_name != None:
        outfile.close()

def usage():
    '''Print a usage message and exit'''
    print('Usage: mkversion [-o out_file] [program [path-to-version.txt]]', file=sys.stderr)
    exit(1)

# Parse the command line.
outfile_name = None
program = 'x3270'
version_txt = './version.txt'
args = sys.argv.copy()[1:]
while len(args) > 0 and args[0].startswith('-'):
    if args[0] == '-o':
        if len(args) < 2:
            usage()
        outfile_name = args[1]
        args = args[1:]
    else:
        usage()
    args = args[1:]

if len(args) > 2:
    usage()
if len(args) > 0:
    program = args[0]
if len(args) > 1:
    version_txt = args[1]

# Run.
mkversion(outfile_name, program, version_txt)